File: 2021/day21.cpp
// https://adventofcode.com/2021/day/21

#include "../common/common.h"
#include "../common/rust.h"

#include <iostream>
#include <map>
#include <print>
#include <ranges>
#include <string>
#include <vector>

namespace stdv = std::views;

using Input = std::pair<u32, u32>;

fn parse(String const& filename) -> Input {
  auto lines = aoc::views::read_lines(filename);
  auto it = std::begin(lines);
  let p1 = aoc::to_number<u32>(
      aoc::trim(*it).substr(sizeof("Player 1 starting position:")));
  ++it;
  let p2 = aoc::to_number<u32>(
      aoc::trim(*it).substr(sizeof("Player 2 starting position:")));
  return {p1, p2};
}

static constexpr u32 WIN_SCORE_PART1 = 1000;
static constexpr u32 WIN_SCORE_PART2 = 21;
static constexpr usize DIRAC_FREQ_OFFSET = 3;

fn next_pos(u32 pos, u32 roll) -> u32 { return (pos + roll - 1) % 10 + 1; }

struct DeterministicDice {
  u32 current{0};
  fn roll() -> u32 {
    current = current % 100 + 1;
    return current;
  }
};

fn solve_case1(Input positions) -> u32 {
  auto [p1, p2] = positions;
  auto score = std::pair<u32, u32>{0, 0};
  auto rolls = u32{};
  auto dice = DeterministicDice{};
  auto roll_dice = [&](u32 pos) -> u32 {
    let rolled = dice.roll() + dice.roll() + dice.roll();
    rolls += 3;
    return next_pos(pos, rolled);
  };
  loop {
    p1 = roll_dice(p1);
    score.first += p1;
    if (score.first >= WIN_SCORE_PART1) {
      return score.second * rolls;
    }
    p2 = roll_dice(p2);
    score.second += p2;
    if (score.second >= WIN_SCORE_PART1) {
      return score.first * rolls;
    }
  }
}

fn dirac_roll_frequency() -> std::array<u32, 7> {
  auto freq = std::array<u32, 7>{};
  for (u32 a = 1; a <= 3; ++a) {
    for (u32 b = 1; b <= 3; ++b) {
      for (u32 c = 1; c <= 3; ++c) {
        freq[a + b + c - DIRAC_FREQ_OFFSET] += 1;
      }
    }
  }
  return freq;
}

struct Universe {
  std::pair<u32, u32> positions;
  std::pair<u32, u32> score;
  bool is_player_one;

  fn operator<=>(Universe const&) const = default;
};

fn solve_case2(Input positions) -> u64 {
  using Multiverse = std::map<Universe, usize>;
  let dirac = dirac_roll_frequency();

  auto multiverse = Multiverse{};
  multiverse[Universe{positions, {0, 0}, true}] = 1;

  auto num_wins = std::pair<u64, u64>{0, 0};

  while (!multiverse.empty()) {
    auto new_multiverse = Multiverse{};
    for (let& [ universe, uni_count ] : multiverse) {
      for (let[roll_idx, freq] : dirac | stdv::enumerate) {
        let roll = static_cast<u32>(roll_idx + DIRAC_FREQ_OFFSET);
        let new_uni_count = uni_count * static_cast<usize>(freq);

        auto new_universe = universe;
        new_universe.is_player_one = !universe.is_player_one;

        if (universe.is_player_one) {
          // Player 1 rolled
          new_universe.positions.first =
              next_pos(universe.positions.first, roll);
          new_universe.score.first += new_universe.positions.first;

          if (new_universe.score.first >= WIN_SCORE_PART2) {
            num_wins.first += new_uni_count;
            continue;
          }
        } else {
          // Player 2 rolled
          new_universe.positions.second =
              next_pos(universe.positions.second, roll);
          new_universe.score.second += new_universe.positions.second;

          if (new_universe.score.second >= WIN_SCORE_PART2) {
            num_wins.second += new_uni_count;
            continue;
          }
        }

        // Nobody won, explore this universe further
        new_multiverse[new_universe] += new_uni_count;
      }
    }
    multiverse = std::move(new_multiverse);
  }

  return std::max(num_wins.first, num_wins.second);
}

int main() {
  std::println("Part 1");
  let example = parse("day21.example");
  AOC_EXPECT_RESULT(739785, solve_case1(example));
  let input = parse("day21.input");
  AOC_EXPECT_RESULT(805932, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(444356092776315ULL, solve_case2(example));
  AOC_EXPECT_RESULT(133029050096658ULL, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2021/src/day21.rs
use std::collections::HashMap;

fn parse(filename: &str) -> (u32, u32) {
    let lines = aoc::file::read_lines(filename);
    let p1 = lines[0].split(": ").nth(1).unwrap().parse().unwrap();
    let p2 = lines[1].split(": ").nth(1).unwrap().parse().unwrap();
    (p1, p2)
}

const WIN_SCORE_PART1: u32 = 1000;
const WIN_SCORE_PART2: u32 = 21;

#[derive(Default)]
struct DeterministicDice {
    current: u32,
}
impl DeterministicDice {
    fn roll(&mut self) -> u32 {
        let result = self.current % 100 + 1;
        self.current = result;
        result
    }
}

fn next_pos(pos: u32, roll: u32) -> u32 {
    (pos + roll - 1) % 10 + 1
}

fn solve_case1((mut p1, mut p2): (u32, u32)) -> u32 {
    let mut score = (0, 0);
    let mut rolls = 0;
    let mut dice = DeterministicDice::default();
    let mut roll_dice = |pos| {
        let rolled = dice.roll() + dice.roll() + dice.roll();
        rolls += 3;
        return next_pos(pos, rolled);
    };
    loop {
        p1 = roll_dice(p1);
        score.0 += p1;
        if (score.0 >= WIN_SCORE_PART1) {
            return score.1 * rolls;
        }
        p2 = roll_dice(p2);
        score.1 += p2;
        if (score.1 >= WIN_SCORE_PART1) {
            return score.0 * rolls;
        }
    }
}

#[derive(Clone, Hash, PartialEq, Eq)]
struct Universe {
    positions: (u32, u32),
    score: (u32, u32),
    is_player_one: bool,
}

fn dirac_rolls() -> impl Iterator<Item = u32> {
    (1..=3).flat_map(|a| (1..=3).flat_map(move |b| (1..=3).map(move |c| a + b + c)))
}

const DIRAC_FREQ_OFFSET: usize = 3;

fn dirac_roll_frequency() -> [u32; 7] {
    let mut freq = [0; 7];
    for roll in dirac_rolls() {
        freq[roll as usize - DIRAC_FREQ_OFFSET] += 1;
    }
    freq
}

fn solve_case2((p1, p2): (u32, u32)) -> usize {
    let universe = Universe {
        positions: (p1, p2),
        score: (0, 0),
        is_player_one: true,
    };

    let mut multiverse: HashMap<Universe, usize> = HashMap::new();
    multiverse.insert(universe, 1);

    let mut num_wins = (0, 0);
    let dirac = dirac_roll_frequency();

    while (!multiverse.is_empty()) {
        let mut new_multiverse = HashMap::new();
        for (universe, uni_count) in multiverse {
            // Roll the dice
            for (roll, num_rolled_universes) in dirac.iter().enumerate() {
                let roll: u32 = (roll + DIRAC_FREQ_OFFSET) as u32;
                let new_uni_count = uni_count * (*num_rolled_universes as usize);

                let mut new_universe = universe.clone();
                new_universe.is_player_one = !universe.is_player_one;

                if (universe.is_player_one) {
                    // Player 1 rolled
                    new_universe.positions.0 = next_pos(universe.positions.0, roll);
                    new_universe.score.0 += new_universe.positions.0;

                    if (new_universe.score.0 >= WIN_SCORE_PART2) {
                        num_wins.0 += new_uni_count;
                        continue;
                    }
                } else {
                    // Player 2 rolled
                    new_universe.positions.1 = next_pos(universe.positions.1, roll);
                    new_universe.score.1 += new_universe.positions.1;

                    if (new_universe.score.1 >= WIN_SCORE_PART2) {
                        num_wins.1 += new_uni_count;
                        continue;
                    }
                }

                // Nobody won, explore this universe further
                *new_multiverse.entry(new_universe).or_default() += new_uni_count;
            }
        }

        multiverse = new_multiverse;
    }

    return num_wins.0.max(num_wins.1);
}

fn main() {
    println!("Part 1");
    let example = parse("day21.example");
    assert_eq!(739785, solve_case1(example));
    let input = parse("day21.input");
    assert_eq!(805932, solve_case1(input));

    println!("Part 2");
    assert_eq!(444356092776315, solve_case2(example));
    assert_eq!(133029050096658, solve_case2(input));
}
